knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/impl/TemplateSnippetProcessor.java

```diff
@@ -44,27 +45,38 @@
 
   private final boolean templateDebug;
 
+
   public TemplateSnippetProcessor(HttpClient httpClient, TemplateEngineConfiguration configuration) {
     this.serviceEngine = new ServiceEngine(httpClient, configuration);
     this.templateDebug = configuration.templateDebug();
   }
 
   public Observable<String> processSnippet(final HtmlFragment fragment, TemplateEngineRequest request) {
-    return fragment.getServices()
+    return Observable.just(fragment)
+        .flatMap(HtmlFragment::getServices)
         .filter(serviceEntry -> serviceEntry.canServeRequest(fragment, request))
         .doOnNext(this::traceService)
```

Comment:
Should we traceService in this moment? I think we should trace service after cache decision.